### ISSUE TITLE
hpctoolkit os=ubuntu18.04 : 'NoneType' object has no attribute 'microarchitecture'

### ISSUE THREAD

**Issue by eugeneswalker** [spack/spack#15306]
UPDATE (alalazo): The issue can be reproduced outside of environments:
```console
$ spack spec hpctoolkit os=ubuntu18.04
Input spec
--------------------------------
hpctoolkit arch=linux-ubuntu18.04-None

Concretized
--------------------------------
==> Error: 'NoneType' object has no attribute 'microarchitecture'
```
and is related to concretization.

---

Using Docker image `spack/ubuntu-bionic` for x86_64 and Spack develop @ fddb3f291cf4 . 

Place the attached [Spack environment file](https://github.com/spack/spack/files/4282866/spack.yaml.txt) into a directory by itself. Rename it to `spack.yaml` and then run the following sequence of commands _in the directory where the spack.yaml is_ in order to reproduce the issue:

**Step 1**
```
$> spack install patchelf
... install succeeds
```

**Step 2**
```
$> spack install
==> Error: 'NoneType' object has no attribute 'microarchitecture'
```

Full stack trace from Step 2:
```
Traceback (most recent call last):
  File "/root/spack/bin/spack", line 64, in <module>
    sys.exit(spack.main.main())
  File "/root/spack/lib/spack/spack/main.py", line 763, in main
    return _invoke_command(command, parser, args, unknown)
  File "/root/spack/lib/spack/spack/main.py", line 488, in _invoke_command
    return_val = command(parser, args)
  File "/root/spack/lib/spack/spack/cmd/install.py", line 273, in install
    concretized_specs = env.concretize()
  File "/root/spack/lib/spack/spack/environment.py", line 948, in concretize
    return self._concretize_separately()
  File "/root/spack/lib/spack/spack/environment.py", line 1016, in _concretize_separately
    concrete = _concretize_from_constraints(uspec_constraints)
  File "/root/spack/lib/spack/spack/environment.py", line 1554, in _concretize_from_constraints
    return s.concretized()
  File "/root/spack/lib/spack/spack/spec.py", line 2298, in concretized
    clone.concretize()
  File "/root/spack/lib/spack/spack/spec.py", line 2148, in concretize
    user_spec_deps=user_spec_deps),
  File "/root/spack/lib/spack/spack/spec.py", line 2626, in normalize
    visited, all_spec_deps, provider_index, tests)
  File "/root/spack/lib/spack/spack/spec.py", line 2552, in _normalize_helper
    dep = self._evaluate_dependency_conditions(dep_name)
  File "/root/spack/lib/spack/spack/spec.py", line 2375, in _evaluate_dependency_conditions
    if self.satisfies(when_spec, strict=True):
  File "/root/spack/lib/spack/spack/spec.py", line 2879, in satisfies
    if not self.architecture.satisfies(other.architecture, strict):
  File "/root/spack/lib/spack/spack/spec.py", line 360, in satisfies
    return self._satisfies_target(other.target, strict=strict)
  File "/root/spack/lib/spack/spack/spec.py", line 383, in _satisfies_target
    min_ok = self_target.microarchitecture >= t_min if t_min else True
AttributeError: 'NoneType' object has no attribute 'microarchitecture'
```

CC @gartung

**Comment by alalazo** [comment]
@eugeneswalker After some effort spent trying to reduce the environment that was posted, I was able to reproduce the same behavior with:
```yaml
spack:
  specs:
  - hpctoolkit os=ubuntu18.04 
  view: false
  concretization: separately
```
In fact the behavior doesn't seem to be related to Spack environments at all:
```console
$ spack spec hpctoolkit os=ubuntu18.04
Input spec
--------------------------------
hpctoolkit arch=linux-ubuntu18.04-None

Concretized
--------------------------------
==> Error: 'NoneType' object has no attribute 'microarchitecture'
```
As a workaround you can specify separately i.e. out of the matrix `hpctoolkit` and omit the offending `os=ubuntu18.04` - though there's definitely an issue with `hpctoolkit` concretization.

**Comment by gartung** [comment]
Using arch=linux-ubuntu18.04-ppc64le works
```
Input spec
--------------------------------
hpctoolkit arch=linux-ubuntu18.04-ppc64le

Concretized
--------------------------------
hpctoolkit@2020.03.01%gcc@7.4.0~all-static~bgq~cray~cuda~mpi~papi arch=linux-ubuntu18.04-ppc64le
    ^binutils@2.33.1%gcc@7.4.0+gold~headers+libiberty~nls~plugins arch=linux-ubuntu18.04-ppc64le
        ^zlib@1.2.11%gcc@7.4.0+optimize+pic+shared arch=linux-ubuntu18.04-ppc64le
    ^boost@1.72.0%gcc@7.4.0+atomic+chrono~clanglibcpp~container~context~coroutine cxxstd=98 +date_time~debug+exception~fiber+filesystem+graph~icu+iostreams+locale+log+math~mpi+multithreaded~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread+timer~versionedlayout visibility=global +wave arch=linux-ubuntu18.04-ppc64le
        ^bzip2@1.0.8%gcc@7.4.0+shared arch=linux-ubuntu18.04-ppc64le
            ^diffutils@3.7%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
                ^libiconv@1.16%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
    ^dyninst@10.1.0%gcc@7.4.0 build_type=RelWithDebInfo +openmp~stat_dysect~static arch=linux-ubuntu18.04-ppc64le
        ^cmake@3.16.2%gcc@7.4.0~doc+ncurses+openssl+ownlibs~qt arch=linux-ubuntu18.04-ppc64le
            ^ncurses@6.1%gcc@7.4.0~symlinks+termlib arch=linux-ubuntu18.04-ppc64le
                ^pkgconf@1.6.3%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
            ^openssl@1.1.1d%gcc@7.4.0+systemcerts arch=linux-ubuntu18.04-ppc64le
                ^perl@5.30.1%gcc@7.4.0+cpanm+shared+threads arch=linux-ubuntu18.04-ppc64le
                    ^gdbm@1.18.1%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
                        ^readline@8.0%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
        ^elfutils@0.178%gcc@7.4.0+bzip2~nls+xz arch=linux-ubuntu18.04-ppc64le
            ^m4@1.4.18%gcc@7.4.0 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-ubuntu18.04-ppc64le
                ^libsigsegv@2.12%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
            ^xz@5.2.4%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
        ^intel-tbb@2020.1%gcc@7.4.0 cxxstd=default patches=62ba015ebd1819c45bef47411540b789b493e31ca668c4ff4cb2afcbc306b476,ce1fb16fb932ce86a82ca87cf0431d1a8c83652af9f552b264213b2ff2945d73 +shared+tm arch=linux-ubuntu18.04-ppc64le
        ^libiberty@2.33.1%gcc@7.4.0+pic arch=linux-ubuntu18.04-ppc64le
    ^libdwarf@20180129%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
    ^libmonitor@2018.07.18%gcc@7.4.0~bgq+hpctoolkit~krellpatch arch=linux-ubuntu18.04-ppc64le
    ^libpfm4@4.10.1%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
    ^libunwind@2018.10.12%gcc@7.4.0+xz arch=linux-ubuntu18.04-ppc64le
        ^autoconf@2.69%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
        ^automake@1.16.1%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
        ^libtool@2.4.6%gcc@7.4.0 arch=linux-ubuntu18.04-ppc64le
    ^mbedtls@2.16.1%gcc@7.4.0 build_type=Release +pic arch=linux-ubuntu18.04-ppc64le
    ^xerces-c@3.2.2%gcc@7.4.0 cxxstd=default transcoder=iconv arch=linux-ubuntu18.04-ppc64le
```

**Comment by alalazo** [comment]
@gartung See the fix in #15319 There is an issue when checking a constraint where:
1. The architecture spec is not concrete (lacks the target)
2. `strict=True`

**Comment by mwkrentel** [comment]
If the problem involves `strict=True`, then this comes from intel-tbb.

```
  if spec.satisfies('@2017.8,2018.1:', strict=True):
```

This was necessary to fix a problem with variable-length version
numbers (eg, 2017 and 2017.1).  See #13989

Hpctoolkit depends directly on intel-tbb and on dyninst which depends
on intel-tbb.  The odd thing is that the concretization succeeds with
intel-tbb and dyninst, but not with hpctoolkit.

```
$ spack spec intel-tbb os=fedora26
Input spec
--------------------------------
intel-tbb arch=linux-fedora26-None

Concretized
--------------------------------
intel-tbb@2020.1%gcc@7.3.1 cxxstd=default patches=62ba015ebd1819c45bef47411540b789b493e31ca668c4ff4cb2afcbc306b476,ce1fb16fb932ce86a82ca87cf0431d1a8c83652af9f552b264213b2ff2945d73 +shared+tm arch=linux-fedora26-westmere
...


$ spack spec dyninst os=fedora26
Input spec
--------------------------------
dyninst arch=linux-fedora26-None

Concretized
--------------------------------
dyninst@10.1.0%gcc@7.3.1 build_type=RelWithDebInfo +openmp~stat_dysect~static arch=linux-fedora26-westmere
...


$ spack spec hpctoolkit os=fedora26
Input spec
--------------------------------
hpctoolkit arch=linux-fedora26-None

Concretized
--------------------------------
==> Error: 'NoneType' object has no attribute 'microarchitecture'
```

So, maybe the problem involves two dependency paths from hpctoolkit to
intel-tbb in the same dag.

**Comment by alalazo** [comment]
@mwkrentel See the fix in #15319, the problem is due to a bug in `ArchSpec.satisfies`.
